src/serial_io.cpp

```diff
@@ -1,228 +1,256 @@
  */
-uint8_t * 
-SerialIO::getAddress(void) 
+void 
+SerialIO::setConfig() 
 {
-  return input_address.bytes;
+  while (board_state == CONFIG) {
+    switch (serial_state) {
+    case FLUSHING:
+      flushSerial();  
+      serial_state = READING;
+      break;
+    
+    case READING:
+      setFromSerial(&input_channel, (uint32_t) CHANNEL_BYTES);
+      setFromSerial(input_address.bytes, (uint32_t) ADDRESS_BYTES);
+      board_state = READY;
+      break;
+    
+    default:
+      break;
+    }
+  }   
 }
 
 /*
- * Getter for channel, set after setTransmission() is run
+ * Function setExtension() reads data sent over serial to set the desired file
+ * extension so we can decode our sent hex file on the RX side.
  */
-uint8_t 
-SerialIO::getChannel(void) 
+void
+SerialIO::setExtension()
 {
-  return input_channel;
+  handShake();
+  setFromSerial(file_extension, (uint32_t) EXTENSION_BYTES, false); 
+
+  // for debugging
+  send(input_channel);
+  send(input_address.num);
+  send((char *)file_extension);
 }
 
 
 /*
+ * Function setFileHexChunk() reads raw hex of our to-be-sent file over serial
+ * to set our next file chunk, which will be sent over from our TX board to our 
+ * RX board.  Note that our file must be sent in chunks because we are unable to 
+ * fit it entirely on the board, so we must determine the size of each chunk via 
+ * setFileHexSize(), sent to us over serial via the sed_hex.py script
+ */
+void 
+SerialIO::setFileHexChunk() 
+{
+
+  setFileHexSize();
+  handShake();
+  setFromSerial(file_chunk, next_chunk_size.num, true);
+
+  // for debugging
+  send(file_chunk);
+}
+
+/*
+ * Function setFileHexSize() reads data sent over serial to set the size of our
+ * next hex chunk of our to-be-sent file.
  */
 void 
 SerialIO::setFileHexSize() 
 {
-  char curr_char;
-  uint8_t sent_bytes {0};
-  
   handShake();
+  setFromSerial(next_chunk_size.bytes, (uint32_t) CUNK_SIZE_BYTES);
+  
+  // for debugging
+  send(next_chunk_size.num);
+}
 
-  while (sent_bytes != CUNK_SIZE_BYTES) {
-    while ((sent_bytes != CUNK_SIZE_BYTES) && Serial.available()) {
-      curr_char = (char) (Serial.read());
-      *p_next_chunk_size = curr_char;
-      p_next_chunk_size++;
-      sent_bytes++;
-      if (sent_bytes == CUNK_SIZE_BYTES) {
-          print(next_chunk_size.num);
-      }
 
-      
+/* -----Auxillary Functions----- */
+
+/*
+ * Function handShake() establishes a line of communication between the computer and Arduino.
+ * First, we (from the Arduino) wait for the computer to say it is ready, then we say we are 
+ * ready back by sending and receiving CONFIRMATION_CHAR over serial
+ */
+void
+SerialIO::handShake()
+{
+  char curr_char {'a'}; // arbirary initiallization != CONFIRMATION_CHAR
+
+  /* stay in while loop until computer says it is ready */
+  while (curr_char != CONFIRMATION_CHAR) {
+    while (curr_char != CONFIRMATION_CHAR && Serial.available()) {
+      curr_char = (char) (Serial.read());
```

Comment:
I've seen this loop a few times.. how can we be sure that it doesn't get stuck here? In the case where the serial is not available and the curr char is not the CONFIRMATION_CHAR, would that not be an infinite loop? Or would that only happen if we like unplugged the board?